libsolidity/formal/SymbolicVariables.h

```diff
@@ -190,77 +192,75 @@ class SymbolicFunctionVariable: public SymbolicVariable
 };
 
 /**
- * Specialization of SymbolicVariable for Mapping
+ * Specialization of SymbolicVariable for Enum
  */
-class SymbolicMappingVariable: public SymbolicVariable
+class SymbolicEnumVariable: public SymbolicVariable
 {
 public:
-	SymbolicMappingVariable(
+	SymbolicEnumVariable(
 		frontend::TypePointer _type,
 		std::string _uniqueName,
 		EncodingContext& _context
 	);
 };
 
 /**
- * Specialization of SymbolicVariable for Array
+ * Specialization of SymbolicVariable for Tuple
  */
-class SymbolicArrayVariable: public SymbolicVariable
+class SymbolicTupleVariable: public SymbolicVariable
 {
 public:
-	SymbolicArrayVariable(
+	SymbolicTupleVariable(
 		frontend::TypePointer _type,
-		frontend::TypePointer _originalTtype,
 		std::string _uniqueName,
 		EncodingContext& _context
 	);
-	SymbolicArrayVariable(
+	SymbolicTupleVariable(
 		SortPointer _sort,
 		std::string _uniqueName,
 		EncodingContext& _context
 	);
 
-	SymbolicArrayVariable(SymbolicArrayVariable&&) = default;
-
-	Expression currentValue(frontend::TypePointer const& _targetType = TypePointer{}) const override;
-};
-
-/**
- * Specialization of SymbolicVariable for Enum
- */
-class SymbolicEnumVariable: public SymbolicVariable
-{
-public:
-	SymbolicEnumVariable(
-		frontend::TypePointer _type,
-		std::string _uniqueName,
-		EncodingContext& _context
+	std::vector<SortPointer> const& components();
+	Expression component(
+		size_t _index,
+		TypePointer _fromType = nullptr,
+		TypePointer _toType = nullptr
 	);
 };
 
 /**
- * Specialization of SymbolicVariable for Tuple
+ * Specialization of SymbolicVariable for Array
  */
-class SymbolicTupleVariable: public SymbolicVariable
+class SymbolicArrayVariable: public SymbolicVariable
 {
 public:
-	SymbolicTupleVariable(
+	SymbolicArrayVariable(
 		frontend::TypePointer _type,
+		frontend::TypePointer _originalTtype,
 		std::string _uniqueName,
 		EncodingContext& _context
 	);
-	SymbolicTupleVariable(
+	SymbolicArrayVariable(
 		SortPointer _sort,
 		std::string _uniqueName,
 		EncodingContext& _context
 	);
 
-	std::vector<SortPointer> const& components();
-	Expression component(
-		size_t _index,
-		TypePointer _fromType = nullptr,
-		TypePointer _toType = nullptr
-	);
+	SymbolicArrayVariable(SymbolicArrayVariable&&) = default;
+
+	Expression currentValue(frontend::TypePointer const& _targetType = TypePointer{}) const override;
+	Expression valueAtIndex(int _index) const override;
+	Expression resetIndex() override { SymbolicVariable::resetIndex(); return m_pair.resetIndex(); }
+	Expression setIndex(unsigned _index) override { SymbolicVariable::setIndex(_index); return m_pair.setIndex(_index); }
+	Expression increaseIndex() override { SymbolicVariable::increaseIndex(); return m_pair.increaseIndex(); }
+	Expression array();
```

Comment:
I would have gone for calling this ``elements()`` or something like that...

Reply:
Calling `array` `elements`? Not sure I get it

Reply:
Just a different name for the function... I mean this thing is a symbolic array - it's a bit weird that it has a member called ``array``... so personally I'd call the function ``elements()`` - i.e. you query the expression corresponding to the array elements, not to the array length...

Then you'd not have ``select(array->array(), index)``, but ``select(array->elements(), index)``, etc...
But that's just a side-node about personal taste - I'm fine with keeping this be called ``array`` :-).

Reply:
Ah, right. I thought you meant the tuple elements. Fine with changing.